src/rgw/rgw_redis_lock.cc

```diff
@@ -0,0 +1,100 @@
+#include "rgw_redis_lock.h"
+
+#include "common/async/blocked_completion.h"
+
+namespace rgw {
+namespace redislock {
+
+struct initiate_exec {
+  std::shared_ptr<boost::redis::connection> conn;
+
+  using executor_type = boost::redis::connection::executor_type;
+  executor_type get_executor() const noexcept { return conn->get_executor();
+  }
+
+  template <typename Handler, typename Response>
+  void operator()(Handler handler, const boost::redis::request& req,
+                  Response& resp) {
+    auto h = boost::asio::consign(std::move(handler), conn);
+    return boost::asio::dispatch(
+        get_executor(), [c = conn, &req, &resp, h = std::move(h)]() mutable {
+          return c->async_exec(req, resp, std::move(h));
+        });
+  }
+};
+
+template <typename Response, typename CompletionToken>
+auto async_exec(std::shared_ptr<connection> conn,
+                const boost::redis::request& req, Response& resp,
+                CompletionToken&& token) {
+  return boost::asio::async_initiate<
+      CompletionToken, void(boost::system::error_code, std::size_t)>(
+      initiate_exec{std::move(conn)}, token, req, resp);
```

Comment:
why calling `std::move()` on the shared pointer?
I think this will leave the shared pointer in the calling code in undefind state

Reply:
in line 64 it is probably ok, because the moved object is not going to be used after the move.
but in line 81 and line 97 it is probaly a problem, because the shared pointer is passed by reference to these funtions, meaning, there is still an object outside of these functions (currently in the test code) that was moved and but may be used after the move